File: android/FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/internal/DCModule.java
package org.firstinspires.ftc.robotcontroller.internal;

import android.app.Activity;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorController;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareDevice;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.configuration.MotorConfigurationType;
import com.qualcomm.robotcore.util.Hardware;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.RegEx;

class TestClass {

}

public class DCModule extends ReactContextBaseJavaModule {

    private static final String TEST_VALUE = "FOO";

    private FtcRobotControllerActivity mainActivity;

   public DCModule(ReactApplicationContext reactContext) {
       super(reactContext);

       this.mainActivity = (FtcRobotControllerActivity) reactContext.getCurrentActivity();
   }

    @Override
    public String getName() {
        return "DCTest";
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("motor", "VALUE");
        return constants;
    }

    @ReactMethod
    public void start() {
       mainActivity.shairedOpMode.runOpMode();
    }

    @ReactMethod
    public void init() {
        mainActivity.shairedOpMode.init();
        mainActivity.shairedOpMode.start();
    }

    @ReactMethod
    public int getDCMotor(String name) {
        return this.mainActivity.shairedOpMode.getDCMotor(name);
    }

    @ReactMethod
    public void setPower(int motorIndex, double power) {
       this.mainActivity.shairedOpMode.setPower(motorIndex, power);
    }

    @ReactMethod
    public String test() {
        return this.mainActivity.shairedOpMode.testValue;
    }
}

File: android/FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/internal/DCModule.kt
package org.firstinspires.ftc.robotcontroller.internal

import android.app.Activity
import android.app.ActivityManager
import android.content.Context
import android.content.ContextWrapper
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.util.HashMap

class DCModule: ReactContextBaseJavaModule {

	// simple value to test constants
	private val TEST_VALUE = "FOO"

	// main activity so that we can get the shared op mode
	private var mainActivity: FtcRobotControllerActivity? = null

	// holding our react context incase we want to update our main activity
	private var mContext: ReactApplicationContext? = null

	constructor(reactContext: ReactApplicationContext): super(reactContext) {
		println("Given context: ")
		println(reactContext)

		// set the main activity from the context we are given
		this.mainActivity = reactContext.currentActivity as FtcRobotControllerActivity?
		this.mContext = reactContext
	}

	override fun getName(): String {
		return "DCTest"
	}

	override fun getConstants(): Map<String, Any>? {
		val constants = HashMap<String, Any>()
		constants["motor"] = "VALUE"
		return constants
	}

	@ReactMethod
	fun getDCMotor(name: String, cb: Callback) {
		cb.invoke(sharedOpMode!!.getDCMotor(name))
	}

	@ReactMethod
	fun setPower(motorIndex: Int, power: Double) {
	   sharedOpMode!!.setPower(motorIndex, power)
	}

	@ReactMethod
	fun test(): String {
		return sharedOpMode!!.testValue
	}

	@ReactMethod
	fun telemetry(message: String) {
	   sharedOpMode!!.telemetry.addData("> ", message)
	   sharedOpMode!!.telemetry.update()
	}
}